python/src/spark_rapids_ml/umap.py

```diff
@@ -1137,26 +1232,54 @@ def _train_udf(pdf_iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
                 )
                 inputs.append((features, label, row_number))
 
+            if cuda_managed_mem_enabled and use_sparse_array:
+                concated_nnz = sum(triplet[0].nnz for triplet in inputs)  # type: ignore
+                if concated_nnz > np.iinfo(np.int32).max:
+                    logger.warn(
+                        "The number of non-zero values of a partition is larger than the int32 index dtype of cupyx csr_matrix."
+                    )
+                else:
+                    inputs = [
+                        (cupyx.scipy.sparse.csr_matrix(row[0]), row[1], row[2])
+                        for row in inputs
+                    ]
+
             # call the cuml fit function
             # *note*: cuml_fit_func may delete components of inputs to free
             # memory.  do not rely on inputs after this call.
             embedding, raw_data = cuml_fit_func(inputs, params).values()
+
             logger.info("Cuml fit complete")
 
             num_sections = (len(embedding) + chunk_size - 1) // chunk_size
 
             for i in range(num_sections):
                 start = i * chunk_size
                 end = min((i + 1) * chunk_size, len(embedding))
-
-                yield pd.DataFrame(
-                    data=[
-                        {
-                            "embedding_": embedding[start:end].tolist(),
-                            "raw_data_": raw_data[start:end].tolist(),
-                        }
-                    ]
-                )
+                if use_sparse_array:
+                    csr = raw_data[start:end]
+                    indices = csr.indices
+                    indptr = csr.indptr
+                    data = csr.data
+                    yield pd.DataFrame(
+                        data=[
+                            {
+                                "embedding_": embedding[start:end].tolist(),
+                                "indices": indices.tolist(),
+                                "indptr": indptr.tolist(),
+                                "data": data.tolist(),
+                            }
+                        ]
+                    )
+                else:
+                    yield pd.DataFrame(
+                        data=[
```

Comment:
The data = [...] is a bug as it is packing everything into a single row so the whole thing is one Arrow record. Also explains why the CSR components with mismatched sizes were passing tests despite mismatched sizes.